### DISCUSSION TITLE
Smallrye config with list

### DISCUSSION THREAD

**Discussion by alex-martel** [quarkusio/quarkus#24107] [Q&A] [answered]
I just got a problem when configuring my application and I really do not know how to get around it. I think it is by design, but I feel this design quite annoying.

I am using common lang Range class. `Range<Integer>` to be precise. I used to have my configuration like this;
`range.valid=00-90`
with the interface as follow;
```
@WithName("range.valid")
@WithConverter(RangeConverter.class)
 Range<Integer> allowedRange();
```

Recently we had the requirement of supporting multiple ranges, so I told myself "oh easy", just have to change the type to a list and it should work almost as is.. wrong
`range.valid=00-90,91=93`
```
@WithName("range.valid")
@WithConverter(RangeConverter.class)
List<Range<Integer>> allowedRange();
```

I was under the impression the **string from the config would still be splitted at the comma**, then the list would be built but instead I get;
`java.lang.IllegalArgumentException: SRCFG00013: No Converter registered for class org.apache.commons.lang3.Range`
which I mean, it does exists, as I did not touch the previous converter at all.

Then I went with creating a subinterface route as the config guide shows; something like this;
```
public interface Config {
   . . . .
    @WithName("range")
    List<Range> allowedRange();
    interface Range {
        @WithConverter(RangeConverter.class)
        @WithName("range.valid")
        org.apache.commons.lang3.Range<Integer> valid();
    }
```

I could get this to work only by using indexed properties like this;
```
range[0].valid=00-90
range[1].valid=92-95
```
this would not work;
range.valid=00-90,92-95
I would get;
`range.valid does not map to any root`

The problem I have is that we use env variables, to configure the app within k8s, and **indexed properties aren't supported this way**. 

So I am blocked at the moment with that seemingly really simple problem. Any help or workarounds would be greatly appreciated.

Thank you!

**Comment by alex-martel** [comment]
In the end I ended up creating my own `Ranges` object with a add method, which contained the `List<Range<Integer>>`. Works but it's cumbersome overall I find versus supporting List (or anything else really via converter) out of the box.

**Comment by radcortez** [comment, marked as answer]
This should work:

```java
@WithConverter(RangeConverter.class)
List<Range> list();
```

But it seems that in the case of the list, the `RangeConverter` is not being added to the converter list, so you get the error message `No Converter registered...`. Please, register the Converter manually like this: https://quarkus.io/guides/config-extending-support#custom-converter.

In the meanwhile, I'll have a look into SmallRye Config to make this work without having to register the Converter with the ServiceLoader in the case of a List.

**Reply by alex-martel** [reply]
Thanks for looking into it.

It does not seem to work. I have added the file
`src/main/resources/META-INF/services/org.eclipse.microprofile.config.spi.Converter`
inside that file I have
`org.acme.converters.config.RangeConverter`

When running the app I still have;
`SRCFG00013: No Converter registered for class org.apache.commons.lang3.Range`

I tried both the original converter `RangeConverter implements Converter<Range<Integer>> ` and this `RangeConverter implements Converter<List<Range<Integer>>>`

**Reply by radcortez** [reply]
Weird.

I've tried this: https://github.com/radcortez/quarkus-playground/commit/aee1d0fd5574bb19527a57c0058e8b9f7403eaa3

And it works for me:

```console
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 6.674 s - in com.radcortez.quarkus.playground.number.config.RangeConfigTest
Mar 16, 2022 9:40:08 PM io.quarkus.opentelemetry.exporter.otlp.runtime.LateBoundBatchSpanProcessor logDelegateNotFound
WARN: No BatchSpanProcessor delegate specified, no action taken.
Mar 16, 2022 9:40:08 PM io.quarkus.bootstrap.runner.Timing printStopTime
INFO: Quarkus stopped in 0.058s
```

**Reply by radcortez** [reply]
Anyway, let me have a look on the SmallRye Config side, so that extra registration is not required.

**Reply by radcortez** [reply]
This should fix it: https://github.com/smallrye/smallrye-config/pull/724

**Reply by alex-martel** [reply]
I did test now with SR Config 2.9.2 and it indeed does work. Thank you.

Side question, should the documentation about manually registering the converter be removed as it does not seem necessary for any converters? I can make the change if you think it is ok.

**Reply by radcortez** [reply]
It is still necessary if you use the programmatic API (and you don't have a mapping with the converter registration).
